File: go/internal/service/user_service.go
package service

import (
	"context"
	"fmt"

	"github.com/hemanthpathath/flex-db/go/internal/repository"
)

// UserService handles user business logic
type UserService struct {
	repo repository.UserRepository
}

// NewUserService creates a new UserService
func NewUserService(repo repository.UserRepository) *UserService {
	return &UserService{repo: repo}
}

// Create creates a new user
func (s *UserService) Create(ctx context.Context, email, displayName string) (*repository.User, error) {
	if email == "" {
		return nil, fmt.Errorf("email is required")
	}
	if displayName == "" {
		return nil, fmt.Errorf("display_name is required")
	}

	user := &repository.User{
		Email:       email,
		DisplayName: displayName,
	}

	return s.repo.Create(ctx, user)
}

// GetByID retrieves a user by ID
func (s *UserService) GetByID(ctx context.Context, id string) (*repository.User, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}
	return s.repo.GetByID(ctx, id)
}

// Update updates an existing user
func (s *UserService) Update(ctx context.Context, id, email, displayName string) (*repository.User, error) {
	if id == "" {
		return nil, fmt.Errorf("id is required")
	}

	user, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}

	if email != "" {
		user.Email = email
	}
	if displayName != "" {
		user.DisplayName = displayName
	}

	return s.repo.Update(ctx, user)
}

// Delete deletes a user
func (s *UserService) Delete(ctx context.Context, id string) error {
	if id == "" {
		return fmt.Errorf("id is required")
	}
	return s.repo.Delete(ctx, id)
}

// List retrieves users with pagination
func (s *UserService) List(ctx context.Context, pageSize int32, pageToken string) ([]*repository.User, *repository.ListResult, error) {
	opts := repository.ListOptions{
		PageSize:  int(pageSize),
		PageToken: pageToken,
	}
	return s.repo.List(ctx, opts)
}

// AddToTenant adds a user to a tenant
func (s *UserService) AddToTenant(ctx context.Context, tenantID, userID, role string) (*repository.TenantUser, error) {
	if tenantID == "" {
		return nil, fmt.Errorf("tenant_id is required")
	}
	if userID == "" {
		return nil, fmt.Errorf("user_id is required")
	}

	tenantUser := &repository.TenantUser{
		TenantID: tenantID,
		UserID:   userID,
		Role:     role,
	}

	return s.repo.AddToTenant(ctx, tenantUser)
}

// RemoveFromTenant removes a user from a tenant
func (s *UserService) RemoveFromTenant(ctx context.Context, tenantID, userID string) error {
	if tenantID == "" {
		return fmt.Errorf("tenant_id is required")
	}
	if userID == "" {
		return fmt.Errorf("user_id is required")
	}
	return s.repo.RemoveFromTenant(ctx, tenantID, userID)
}

// ListTenantUsers lists users in a tenant
func (s *UserService) ListTenantUsers(ctx context.Context, tenantID string, pageSize int32, pageToken string) ([]*repository.TenantUser, *repository.ListResult, error) {
	if tenantID == "" {
		return nil, nil, fmt.Errorf("tenant_id is required")
	}

	opts := repository.ListOptions{
		PageSize:  int(pageSize),
		PageToken: pageToken,
	}
	return s.repo.ListTenantUsers(ctx, tenantID, opts)
}

File: app/service/user_service.py
"""
User service implementation.
"""

from typing import List, Tuple

from app.repository import User, TenantUser, UserRepository, ListOptions, ListResult


class UserService:
    """User business logic service."""

    def __init__(self, repo: UserRepository):
        self.repo = repo

    async def create(self, email: str, display_name: str) -> User:
        """Create a new user."""
        if not email:
            raise ValueError("email is required")
        if not display_name:
            raise ValueError("display_name is required")

        user = User(email=email, display_name=display_name)
        return await self.repo.create(user)

    async def get_by_id(self, id: str) -> User:
        """Retrieve a user by ID."""
        if not id:
            raise ValueError("id is required")
        return await self.repo.get_by_id(id)

    async def update(self, id: str, email: str, display_name: str) -> User:
        """Update an existing user."""
        if not id:
            raise ValueError("id is required")

        user = await self.repo.get_by_id(id)

        if email:
            user.email = email
        if display_name:
            user.display_name = display_name

        return await self.repo.update(user)

    async def delete(self, id: str) -> None:
        """Delete a user."""
        if not id:
            raise ValueError("id is required")
        await self.repo.delete(id)

    async def list(self, page_size: int, page_token: str) -> Tuple[List[User], ListResult]:
        """Retrieve users with pagination."""
        opts = ListOptions(page_size=page_size, page_token=page_token)
        return await self.repo.list(opts)

    async def add_to_tenant(self, tenant_id: str, user_id: str, role: str) -> TenantUser:
        """Add a user to a tenant."""
        if not tenant_id:
            raise ValueError("tenant_id is required")
        if not user_id:
            raise ValueError("user_id is required")

        tenant_user = TenantUser(tenant_id=tenant_id, user_id=user_id, role=role)
        return await self.repo.add_to_tenant(tenant_user)

    async def remove_from_tenant(self, tenant_id: str, user_id: str) -> None:
        """Remove a user from a tenant."""
        if not tenant_id:
            raise ValueError("tenant_id is required")
        if not user_id:
            raise ValueError("user_id is required")
        await self.repo.remove_from_tenant(tenant_id, user_id)

    async def list_tenant_users(self, tenant_id: str, page_size: int, page_token: str) -> Tuple[List[TenantUser], ListResult]:
        """List users in a tenant."""
        if not tenant_id:
            raise ValueError("tenant_id is required")

        opts = ListOptions(page_size=page_size, page_token=page_token)
        return await self.repo.list_tenant_users(tenant_id, opts)
